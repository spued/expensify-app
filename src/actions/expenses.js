import { v1 as uuidv1 } from 'uuid';
import db from '../firebase/firebase'

export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
})

export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => {
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };
        return db.ref('expenses').push(expense).then((ref) => {
            dispatch(addExpense({
                id : ref.key,
                ...expense
            }));
        })
    }
}

export const removeExpense = ({id} = {}) => ({
    type : 'REMOVE_EXPENSE',
    id
})

export const editExpense = (id, updates) => ({
    type : 'EDIT_EXPENSE',
    id,
    updates
})

// Set expenses
export const setExpenses = (expenses) => ({
        type: 'SET_EXPENSES',
        expenses
});

export const startSetExpenses = () => {
    return (dispatch) => {
        return db.ref('expenses').once('value').then((snapshot) => {
            let _expenses = [];
            snapshot.forEach((childSnapshot) => {
                //console.log(childSnapshot.val());
                _expenses.push({
                    id : parseInt(childSnapshot.key),
                    ...childSnapshot.val()
                });
            });
            dispatch(setExpenses(_expenses))
        });
        
    };
};