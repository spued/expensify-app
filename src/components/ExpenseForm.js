import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


export default class ExpenseForm extends React.Component {
    constructor(props){
        super(props);
        this.state = {
            description : props.expense ? props.expense.description : '',
            note : props.expense ? props.expense.note : '',
            amount : props.expense ? (props.expense.amount/100).toString() : '',
            createdAt : props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused : false,
            error_message : ''
        };
    }
    
    onDescriptionChange = (evt) => {
        const description = evt.target.value;
        this.setState (() => ({ description }));
    };
    onNoteChange = (evt) => {
        const note = evt.target.value;
        this.setState (() => ({ note }));
    };
    onAmountChange = (evt) => {
        const amount = evt.target.value;
        if(amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState (() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState (() => ({ createdAt }));
        }
    };
    onFocusChanged = ({ focused }) => {
        this.setState (() => ({ calendarFocused : focused }));
    };
    onSubmit = (evt) => {
        evt.preventDefault();
        if(!this.state.description || !this.state.amount) {
            this.setState (() => ({ error_message : 'I need description and amount.' }));
        } else {
            this.setState (() => ({ error_message : '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount,10),
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    }
    onRemove = () => {
        this.props.onRemove();
    }
    render() {
        return (
            <div>
                {this.state.error_message && <p>{this.state.error_message}</p>}
                <form onSubmit={this.onSubmit}> 
                    <input 
                        type='text'
                        autoFocus
                        placeholder='Description'
                        value={this.state.description}
                        onChange={this.onDescriptionChange}>
                    </input>
                    <input 
                        type='number'
                        placeholder='Amount'
                        value={this.state.amount}
                        onChange={this.onAmountChange}>
                    </input>
                    <SingleDatePicker
                        date = {this.state.createdAt}
                        onDateChange = {this.onDateChange}
                        focused = { this.state.calendarFocused }
                        onFocusChange = { this.onFocusChanged }
                        numberOfMonths = {1}
                        isOutsideRange= {() => false }
                    />
                    <textarea
                        placeholder='Add the note for your expense. (Optional)'
                        value={this.state.note}
                        onChange={this.onNoteChange}>
                    </textarea>
                    <button>Add Expense</button>
                </form>
                <button id='remove_button' onClick={this.onRemove}>Remove</button>
            </div>
        )
    };      
}
