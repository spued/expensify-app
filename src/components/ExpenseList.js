import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

export const ExpenseList = (props) => (
    <div>
       {
            props.expenses.length === 0 ? (
                <p> No expense.</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={ expense.id } { ...expense }/>;
                }) 
            )
        }
    </div>
);

const mapStateToProps = (state) => {
    let _expenses = selectExpenses(state.expenses, state.filters);
    return {
        expenses: _expenses
    }
}

export default connect(mapStateToProps)(ExpenseList);