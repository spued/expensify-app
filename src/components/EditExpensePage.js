import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

// make it class
export class EditExpensePage extends React.Component { 
    onSubmit = (expense) => {
        //console.log(expense);
        this.props.editExpense( this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemove = () => {
        this.props.removeExpense({ id : this.props.expense.id});
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                This is edit expense component id = { this.props.expense.id}
                <ExpenseForm 
                    expense = { this.props.expense }
                    onSubmit = { this.onSubmit }
                />
                <button onClick={this.onRemove}></button>
            </div>
        )
    }
}
const mapDispatchToProps = (dispatch, props) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense : (id) => dispatch(removeExpense({ id }))
})
const mapStateToProps = (state, props) => ({
        expense: state.expenses.find((expense) => expense.id === props.match.params.id)
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);