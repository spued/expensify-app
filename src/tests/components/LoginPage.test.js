import React from "react";
import { shallow } from "enzyme";
import { LoginPage } from "../../components/LoginPage"
import { startLogin } from "../../actions/auth";

test ('Should show login page', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
})

test ("Should call startLogout on button click", () => {
    const startLogin = jest.fn();
    const wrapper = shallow(<LoginPage startLogin={ startLogin()} />);
    wrapper.find('button').simulate('Click');
    expect(startLogin).toHaveBeenCalled();
})