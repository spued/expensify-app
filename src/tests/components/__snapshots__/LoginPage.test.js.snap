// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should show login page 1`] = `
<div>
  <button>
    Login
  </button>
</div>
`;
